tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/OptimizedSlidingTimeWindowMovingAverages.java

```diff
@@ -0,0 +1,231 @@
+ * sliding time window, every event having the same weight. Unlike the
+ * the more widely used {@link ExponentialMovingAverages} implementation,
+ * with this class the moving average rate drops immediately to zero if the last
+ * marked event is older than the time window.
+ * <p>
+ * A {@link Meter} with {@link com.codahale.metrics.SlidingTimeWindowMovingAverages} works similarly to
+ * a {@link Histogram} with an {@link SlidingTimeWindowArrayReservoir}, but as a Meter
+ * needs to keep track only of the count of events (not the events itself), the memory
+ * overhead is much smaller. SlidingTimeWindowMovingAverages uses buckets with just one
+ * counter to accumulate the number of events (one bucket per seconds, giving 900 buckets
+ * for the 15 minutes time window).
+ */
+public final class OptimizedSlidingTimeWindowMovingAverages implements MovingAverages {
+
+    private static final long TIME_WINDOW_DURATION_MINUTES = 15;
+    private static final long TICK_INTERVAL = TimeUnit.SECONDS.toNanos(1);
+    private static final Duration TIME_WINDOW_DURATION = Duration.ofMinutes(TIME_WINDOW_DURATION_MINUTES);
+
+    // package private for the benefit of the unit test
+    static final int NUMBER_OF_BUCKETS = (int) (TIME_WINDOW_DURATION.toNanos() / TICK_INTERVAL);
+
+    private final AtomicLong lastTick;
+    private final Clock clock;
+
+    /**
+     * One counter per time bucket/slot (i.e. per second, see TICK_INTERVAL) for the entire
+     * time window (i.e. 15 minutes, see TIME_WINDOW_DURATION_MINUTES).
+     */
+    private final List<LongAdder> buckets;
+
+    /**
+     * Index into buckets, pointing at the bucket containing the oldest counts.
+     */
+    private int oldestBucketIndex;
+
+    /**
+     * Index into buckets, pointing at the bucket with the count for the current time (tick).
+     */
+    private int currentBucketIndex;
+
+    /**
+     * Instant at creation time of the time window. Used to calculate the currentBucketIndex
+     * for the instant of a given tick (instant modulo time window duration).
+     */
+    private final Instant bucketBaseTime;
+
+    /**
+     * Instant of the bucket with index oldestBucketIndex.
+     */
+    private Instant oldestBucketTime;
+
+    /**
+     * Creates a new {@link com.codahale.metrics.SlidingTimeWindowMovingAverages}.
+     */
+    OptimizedSlidingTimeWindowMovingAverages() {
+        this(Clock.defaultClock());
+    }
+
+    /**
+     * Creates a new {@link com.codahale.metrics.SlidingTimeWindowMovingAverages}.
+     *
+     * @param clock the clock to use for the meter ticks
+     */
+    OptimizedSlidingTimeWindowMovingAverages(Clock clock) {
+        this.clock = clock;
+        final long startTime = clock.getTick();
+        this.lastTick = new AtomicLong(startTime);
+
+        this.buckets = IntStream.range(0, NUMBER_OF_BUCKETS)
+                .mapToObj(_i -> new LongAdder())
+                .toList();
```

Comment:
note this bit is different than upstream as we can assume JDK 17+ to use unmodifiable list here.